File: app/src/main/java/com/paradoxo/amadeus/util/voz/VozParaTexto.java
/*
 * Created by Carlos Mion On 4 aug 2017
 * Link original http://www.decom.ufop.br/imobilis/tutorial-android-speech-to-text-sem-o-pop-up/
 * Modified by Junior Obom 27 jun 2019
 */

package com.paradoxo.amadeus.util.voz;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.speech.RecognitionListener;
import android.speech.RecognizerIntent;
import android.speech.SpeechRecognizer;
import android.util.Log;

import com.paradoxo.amadeus.modelo.EventoMensagem;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.Locale;

public class VozParaTexto implements RecognitionListener {
    Intent recognizerIntent;
    SpeechRecognizer speechRecognizer;
    public final BackgroundVoiceListener backgroundVoiceListener;

    public static final String ERRO_CORRESPONDENCIA = "Sem correspondência";
    public static final String ERRO_AO_TENTAR_ESCUTAR = " ao tentar escutar";
    public static final String ERRO_TEMPO_ESGOTADO = "Tempo para conexão esgotado";
    public static final String ERRO_AO_ESCUTAR = "Ocorreu um erro ao textar escutar: ";
    public static final String ERRO_PERMISSOES_INSUFICIENTES = "Permissões insuficientes";
    public static final String ERRO_TEMPO_PARA_FALA_ESGOTADO = "Tempo para fala esgotado";
    public static final String ERRO_RECONHECEDOR_DE_VOZ_OCUPADO = "Reconhecedor de voz ocupado";

    public VozParaTexto(Context context) {
        backgroundVoiceListener = new BackgroundVoiceListener();
        speechRecognizer = SpeechRecognizer.createSpeechRecognizer(context);
        speechRecognizer.setRecognitionListener(this);

        recognizerIntent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        recognizerIntent.putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true);
        recognizerIntent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_PREFERENCE, Locale.getDefault());
        recognizerIntent.putExtra(RecognizerIntent.EXTRA_CALLING_PACKAGE, context.getPackageName());
        recognizerIntent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_WEB_SEARCH);

/*
        recognizerIntent.putExtra(RecognizerIntent.EXTRA_SPEECH_INPUT_MINIMUM_LENGTH_MILLIS, 5000);
        Essa flag parece não sutir efeito se implementada assim, uma possível solução seja criar uma
        classe extentendo RecognizerIntent e fazer a modificação diretamente, por hora a
        solução atual basta e isso pode ser revisto no futuro
 */

    }

    @Override
    public void onReadyForSpeech(Bundle bundle) {
    }

    @Override
    public void onBeginningOfSpeech() {
    }

    @Override
    public void onRmsChanged(float v) {
    }

    @Override
    public void onPartialResults(Bundle partialResults) {
        StringBuilder textoEscutado = new StringBuilder();
        ArrayList<String> matches = partialResults.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);

        if (matches != null)
            for (String result : matches)
                textoEscutado.append(result);
                Log.e("Texto parcial", String.valueOf(textoEscutado));

        Log.e("TAG", "Texto esctuado: " + textoEscutado);

        EventBus.getDefault().post(new EventoMensagem(textoEscutado.toString()));
    }

    @Override
    public void onEndOfSpeech() {
    }

    @Override
    public void onError(int error) {
        String nomeErro = "Erro: " + error + ERRO_AO_TENTAR_ESCUTAR;
        switch (error) {
            case SpeechRecognizer.ERROR_NETWORK_TIMEOUT:
                nomeErro = ERRO_TEMPO_ESGOTADO;
                break;
            case SpeechRecognizer.ERROR_SPEECH_TIMEOUT:
                nomeErro = ERRO_TEMPO_PARA_FALA_ESGOTADO;
                break;
            case SpeechRecognizer.ERROR_NO_MATCH:
                nomeErro = ERRO_CORRESPONDENCIA;
                break;
            case SpeechRecognizer.ERROR_RECOGNIZER_BUSY:
                nomeErro = ERRO_RECONHECEDOR_DE_VOZ_OCUPADO;
                break;
            case SpeechRecognizer.ERROR_INSUFFICIENT_PERMISSIONS:
                nomeErro = ERRO_PERMISSOES_INSUFICIENTES;
                break;

        }
        Log.e("TAG", "Erro: " + error + " " + nomeErro);

        EventBus.getDefault().post(new EventoMensagem(ERRO_AO_ESCUTAR, error, nomeErro));
    }

    @Override
    public void onResults(Bundle bundle) {
        Log.e("Resultados da escuta", "Terminou de gravar");

        ArrayList<String> resultados = bundle.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
        String textoEscutado = resultados!=null?  resultados.get(0) : "";
        EventBus.getDefault().post(new EventoMensagem(textoEscutado,true));
    }

    @Override
    public void onEvent(int i, Bundle bundle) {
    }

    @Override
    public void onBufferReceived(byte[] bytes) {
    }

    public class BackgroundVoiceListener {
        public void run() {
            try {
                speechRecognizer.startListening(recognizerIntent);
                Log.e("TAG", "Escutando");
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        public void interrupt() {
            try {
                //speechRecognizer.stopListening();
                speechRecognizer.destroy();
                Log.e("TAG", "Parou de escutar");

            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: app/src/main/java/com/paradoxo/amadeus/util/voz/VozParaTexto.kt
/*
 * Created by Carlos Mion On 4 aug 2017
 * Link original http://www.decom.ufop.br/imobilis/tutorial-android-speech-to-text-sem-o-pop-up/
 * Modified by Junior Obom 27 jun 2019
 */

package com.paradoxo.amadeus.util.voz

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import android.util.Log
import com.paradoxo.amadeus.modelo.EventoMensagem
import org.greenrobot.eventbus.EventBus
import java.util.Locale

class VozParaTexto(context: Context) : RecognitionListener {
    private val speechRecognizer = SpeechRecognizer.createSpeechRecognizer(context).apply {
        setRecognitionListener(this@VozParaTexto)
    }
    private val recognizerIntent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
        putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true)
        putExtra(RecognizerIntent.EXTRA_LANGUAGE_PREFERENCE, Locale.getDefault())
        putExtra(RecognizerIntent.EXTRA_CALLING_PACKAGE, context.packageName)
        putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_WEB_SEARCH)
    }

    val backgroundVoiceListener = BackgroundVoiceListener()

    companion object {
        const val ERRO_CORRESPONDENCIA = "Sem correspondência"
        const val ERRO_AO_TENTAR_ESCUTAR = " ao tentar escutar"
        const val ERRO_TEMPO_ESGOTADO = "Tempo para conexão esgotado"
        const val ERRO_AO_ESCUTAR = "Ocorreu um erro ao textar escutar: "
        const val ERRO_PERMISSOES_INSUFICIENTES = "Permissões insuficientes"
        const val ERRO_TEMPO_PARA_FALA_ESGOTADO = "Tempo para fala esgotado"
        const val ERRO_RECONHECEDOR_DE_VOZ_OCUPADO = "Reconhecedor de voz ocupado"
    }

    override fun onReadyForSpeech(bundle: Bundle) {}
    override fun onBeginningOfSpeech() {}
    override fun onRmsChanged(v: Float) {}
    override fun onEndOfSpeech() {}
    override fun onEvent(i: Int, bundle: Bundle) {}
    override fun onBufferReceived(bytes: ByteArray) {}

    override fun onPartialResults(partialResults: Bundle) {
        val matches = partialResults.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
        val textoEscutado = matches?.joinToString("") ?: ""
        Log.e("Texto parcial", textoEscutado)
        Log.e("TAG", "Texto esctuado: $textoEscutado")
        EventBus.getDefault().post(EventoMensagem(textoEscutado))
    }

    override fun onError(error: Int) {
        val nomeErro = when (error) {
            SpeechRecognizer.ERROR_NETWORK_TIMEOUT -> ERRO_TEMPO_ESGOTADO
            SpeechRecognizer.ERROR_SPEECH_TIMEOUT -> ERRO_TEMPO_PARA_FALA_ESGOTADO
            SpeechRecognizer.ERROR_NO_MATCH -> ERRO_CORRESPONDENCIA
            SpeechRecognizer.ERROR_RECOGNIZER_BUSY -> ERRO_RECONHECEDOR_DE_VOZ_OCUPADO
            SpeechRecognizer.ERROR_INSUFFICIENT_PERMISSIONS -> ERRO_PERMISSOES_INSUFICIENTES
            else -> "Erro: $error$ERRO_AO_TENTAR_ESCUTAR"
        }
        Log.e("TAG", "Erro: $error $nomeErro")
        EventBus.getDefault().post(EventoMensagem(ERRO_AO_ESCUTAR, error, nomeErro))
    }

    override fun onResults(bundle: Bundle) {
        Log.e("Resultados da escuta", "Terminou de gravar")
        val resultados = bundle.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
        val textoEscutado = resultados?.get(0) ?: ""
        EventBus.getDefault().post(EventoMensagem(textoEscutado, jaTerminou = true))
    }

    inner class BackgroundVoiceListener {
        fun run() {
            try {
                speechRecognizer.startListening(recognizerIntent)
                Log.e("TAG", "Escutando")
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        fun interrupt() {
            try {
                speechRecognizer.destroy()
                Log.e("TAG", "Parou de escutar")
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}
